README.rst
============
gdal2mbtiles
============

Convert GDAL-readable datasets into an MBTiles file
===================================================

**gdal2mbtiles** helps you generate web mapping tiles that can be shown
through a browser-based mapping library on your website.

`GDAL-readable files`_ are images that are georeference, that means that
they are positioned and projected on to the world. In order to display a
dynamic map on the web, you don't want to serve the whole image at once,
so it must be sliced into tiles that are hosted by a tile server.

The MBTiles_ file format was developed by MapBox_ to make tile storage
easier. You can upload the final file to their service, or run your own
tile server. MapBox provides one called TileStream_.

Later versions of GDAL (>= 2) allow generation of mbtiles files via the
``gdal_translate`` and ``gdaladdo`` commands.  However, gdal2mbtiles offers some
advantages:

*  allows you to specify an upper resolution/zoom level.  GDAL always uses the
   native resolution of the input raster to determine the highest zoom level of
   the mbtiles output, whereas gdal2mbtiles can also upsample to create zoom levels
   at a higher resolution than your original file.
* the ``gdal_translate`` command only converts the geotiff at the native resolution,
  so the lower resolutions are added to the file via overviews (``gdaladdo``)
* ``gdaladdo`` can only add overviews down to the zoom level corresponding to
  the size of the tile/block size (256x256).  gdal2mbtiles can always create images
  down to zoom level 1.
* performance: gdal2mbtiles uses pyvips for image processing, which is parallel
  and quick.  Compared to the equivalent processing with GDAL, gdal2mbtiles is
  typically 2-4 times quicker.  For example:

  * a resolution 14 file, 13000x11000 pixels, min resolution 0, max resolution
    14: ~5 minutes with gdal2mbtiles and ~8 minutes with GDAL commands.
  * a resoluton 11 file, 200,000x200,000, zoom level 11 only: ~30min with
    gdal2mbtiles and ~133min with GDAL (with ``GDAL_CACHE_MAX`` and
    ``GDAL_NUM_THREADS`` options)


Installation
============

PyPI package page: https://pypi.python.org/pypi/gdal2mbtiles/

.. warning:: gdal2mbtiles requires Python 2.7 or higher and relies on
  installing the items from the `External Dependencies`_ section below *before*
  the python package.

Using pip::

    $ pip install gdal2mbtiles

From source::

    $ git clone https://github.com/ecometrica/gdal2mbtiles.git
    $ cd gdal2mbtiles
    $ python setup.py install

External Dependencies
---------------------

We rely on GDAL_ to read georeferenced datasets.
Under Debian or Ubuntu, you can install the GDAL library & binary via apt.

Default GDAL versions in Ubuntu LTS:

* Xenial: 1.11
* Bionic: 2.2
* Focal: 3.0

.. warning::
  GDAL 2 is the current supported version.
  GDAL 3 support is in progress - `contributions <#contributing>`_ welcome!

We recommend using the `UbuntuGIS`_ PPA to get more recent versions of GDAL, if
needed, as is the case for Xenial.

.. code-block:: sh

    sudo add-apt-repository ppa:ubuntugis/ppa && sudo apt-get update
    sudo apt-get install gdal-bin libgdal-dev

The ubuntugis PPA also usually includes ``python-gdal`` or ``python3-gdal``
that will install the python bindings at the system level. Installing
that may be enough for you if you aren't planning to use a non-default python
or a `virtual environment`_.

Otherwise, you will also need to install the GDAL python bindings package from
`PyPI <GDAL_PyPI>`_. Make sure to install the version that matches the installed
GDAL library. You can double-check that version with ``gdal-config --version``.

.. code-block:: sh

    pip install \
      --global-option=build_ext \
      --global-option=--gdal-config=/usr/bin/gdal-config \
      --global-option=--include-dirs=/usr/include/gdal/ \
      GDAL=="$(gdal-config --version)"

We also rely on VIPS_ (version 8.2+) to do fast image processing.

Under Debian or Ubuntu, run the following to install it without the GUI nip2::

    $ sudo apt-get install --no-install-recommends libvips libvips-dev

You'll also need a few other libraries to deal with large TIFF files and
to optimize the resulting PNG tiles.

Under Debian or Ubuntu, run the following to install them::

    $ sudo apt-get install libtiff5 optipng pngquant


Command Line Interface
======================

.. code-block:: console

    $ gdal2mbtiles --help
    usage: gdal2mbtiles [-h] [-v] [--name NAME] [--description DESCRIPTION]
                        [--layer-type {baselayer,overlay}] [--version VERSION]
                        [--format {jpg,png}]
                        [--spatial-reference SPATIAL_REFERENCE]
                        [--resampling {near,bilinear,cubic,cubicspline,lanczos}]
                        [--min-resolution MIN_RESOLUTION]
                        [--max-resolution MAX_RESOLUTION] [--fill-borders]
                        [--no-fill-borders] [--zoom-offset N]
                        [--coloring {gradient,palette,exact}]
                        [--color BAND-VALUE:HTML-COLOR]
                        [--colorize-band COLORIZE-BAND]
                        [--png8 PNG8]
                        [INPUT] [OUTPUT]

    Converts a GDAL-readable into an MBTiles file

    optional arguments:
      -h, --help            show this help message and exit
      -v, --verbose         explain what is being done

    Positional arguments:
      INPUT                 GDAL-readable file.
      OUTPUT                Output filename. Defaults to INPUT.mbtiles

    MBTiles metadata arguments:
      --name NAME           Human-readable name of the tileset. Defaults to INPUT
      --description DESCRIPTION
                            Description of the layer. Defaults to ""
      --layer-type {baselayer,overlay}
                            Type of layer. Defaults to "overlay"
      --version VERSION     Version of the tileset. Defaults to "1.0.0"
      --format {jpg,png}    Tile image format. Defaults to "png"

    GDAL warp arguments:
      --spatial-reference SPATIAL_REFERENCE
                            Destination EPSG spatial reference. Defaults to 3857
      --resampling {near,bilinear,cubic,cubicspline,lanczos}
                            Resampling algorithm for warping. Defaults to "near"
                            (nearest-neighbour)

    Rendering arguments:
      --min-resolution MIN_RESOLUTION
                            Minimum resolution/zoom level to render and slice.
                            Defaults to None (do not downsample)
      --max-resolution MAX_RESOLUTION
                            Maximum resolution/zoom level to render and slice.
                            Defaults to None (do not upsample)
      --fill-borders        Fill image to whole world with empty tiles. Default.
      --no-fill-borders     Do not add borders to fill image.
      --zoom-offset N       Offset zoom level by N to fit unprojected images to
                            square maps. Defaults to 0.
      --png8                Quantizes 32-bit RGBA to 8-bit RGBA paletted PNGs.
                            value range from 2 to 256. Default to False.

    Coloring arguments:
      --coloring {gradient,palette,exact}
                            Coloring algorithm.
      --color BAND-VALUE:HTML-COLOR
                            Examples: --color="0:#ff00ff" --color=255:red
      --colorize-band COLORIZE-BAND
                            Raster band to colorize. Defaults to 1


Contributing
============

Reporting bugs and submitting patches
-------------------------------------

Please check our `issue tracker`_ for known bugs and feature requests.

We accept pull requests for fixes and new features.

Development and Testing
-----------------------

We use `Tox`_ and `Pytest`_ to test locally and `CircleCI`_ for remote testing.

1. Clone the repo
2. Install whichever `External Dependencies`_ are suitable for your OS/VM.
3. Create and activate a `virtual environment`_
4. Install tox: ``pip install tox``
5. Set the GDAL_CONFIG env var for tox via the venv activations script.

   If using virtualenv:
   ``echo 'export GDAL_VERSION=$(gdal-config --version)' >> $VIRTUAL_ENV/bin/postactivate``

   If using venv:
   ``echo 'export GDAL_VERSION=$(gdal-config --version)' >> $VIRTUAL_ENV/bin/activate``

6. Run tests to confirm all is working: ``tox``
7. Do some development:

   - Make some changes
   - Run the tests
   - Fix any errors
   - Run the tests again
   - Update CHANGELOG.rst with a line about the change in the UNRELEASED section
   - Add yourself to AUTHORS.rst if not already there
   - Write a nice commit message
   - Repeat

8. Make a PR

You don't need to worry initially about testing in every combination of GDAL
and Ubuntu, leave that to the remote CI build matrix when you make a PR and let
the reviewers figure out if it needs more work from that.

Credits
=======

Maxime Dupuis and Simon Law wrote this program, with the generous
support of Ecometrica_.

See AUTHORS.rst for the full list of contributors.

.. _GDAL-readable files: http://www.gdal.org/formats_list.html
.. _MBTiles: http://mapbox.com/developers/mbtiles/
.. _MapBox: http://mapbox.com/
.. _TileStream: https://github.com/mapbox/tilestream

.. _GDAL: http://www.gdal.org/
.. _UbuntuGIS: https://launchpad.net/~ubuntugis/
.. _VIPS: http://www.vips.ecs.soton.ac.uk/

.. _GDAL_PyPI: https://https://pypi.org/project/GDAL/
.. _Tox: https://tox.readthedocs.io/
.. _Pytest: https://docs.pytest.org/
.. _virtual environment: https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/#creating-a-virtual-environment

.. _issue tracker: https://github.com/ecometrica/gdal2mbtiles/issues
.. _Ecometrica: http://ecometrica.com/

.. _CircleCI: https://circleci.com/


CHANGELOG.rst
=============
Release Notes
=============

The format is based on `Keep a Changelog <https://keepachangelog.com/en/1.0.0/>`_
and this project attempts to adhere to `Semantic Versioning <https://semver.org/spec/v2.0.0.html>`_.

Unreleased
------------

2.1.5
------
* Fix SpatialReference.GetEPSGCode failling to recognise QGIS style PROJCS name.

2.1.4
------
* Fixing GDAL 3 backwards incompatible change which switches axis in coordinate transformation - see: https://github.com/OSGeo/gdal/issues/1546

2.1.3
------

* Fix float overflow bug
* Unpin pyvips and fix related issue - install pyvips==2.1.8 if any issues
* Fix renderer tests
* Fix deprecation warnings
* Fix python3.7+ pytest errors
* Update author email
* Update CI
* Update README

2.1.2
-----

* Update docs
* Pin pyvips

2.1.1
-----

* Revert commit f7fde54, which reintroduced tiling issues fixed by 9231133.


2.1.0
-----

* Add --png8 argument to quantize 32-bit RGBA to 8-bit RGBA paletted PNGs.
* Specify `NUM_THREADS` option for gdal_translate to use all CPUs
* Update MANIFEST.in to include required files


2.0.0
-----

* Add support for Python 3.5 and 3.6
* Replace vipsCC with pyvips
* Use pytest and tox for testing
* types.py has been renamed to gd_types.py to avoid import conflicts
* Remove multiprocessing as it creates noise in mbtiles output files with
pyvips and doesn't appear to have a significant impact on processing speeds


gdal2mbtiles/main.py
# -*- coding: utf-8 -*-

# Quickstart
# ----------
#
# To turn any GDAL-readable file into an MBTiles file, run:
#   $ gdal2mbtiles filename.tiff
# This creates a filename.mbtiles that can be served from a TMS service like
# Mapbox.
#
# You can explicitly specify an output filename:
#   $ gdal2mbtiles input.tiff output.mbtiles
#
# You can also pipe in any GDAL-readable file:
#   $ cat input.tiff | gdal2mbtiles > output.mbtiles
#
# Licensed to Ecometrica under one or more contributor license
# agreements.  See the NOTICE file distributed with this work
# for additional information regarding copyright ownership.
# Ecometrica licenses this file to you under the Apache
# License, Version 2.0 (the "License"); you may not use this
# file except in compliance with the License.  You may obtain a
# copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import (absolute_import, division, print_function,
                        unicode_literals)


import argparse
from contextlib import contextmanager
import logging
import os
from shutil import copyfileobj
import sys
from tempfile import NamedTemporaryFile

if __name__ == '__main__' and __package__ is None:
    # HACK: Force this to work when called directly
    import gdal2mbtiles
    __package__ = gdal2mbtiles.__name__

from .gdal import RESAMPLING_METHODS, SpatialReference
from .gd_types import rgba
from .mbtiles import Metadata


COLORING_METHODS = {
    'exact': 'ColorExact',
    'gradient': 'ColorGradient',
    'palette': 'ColorPalette',
}


def coloring_arg(s):
    """Validates --coloring"""
    from gdal2mbtiles import vips
    if s is None:
        return None
    return getattr(vips, COLORING_METHODS[s])


def color_arg(s):
    """Validates --color"""
    try:
        band_value, html_color = s.split(':', 1)
    except ValueError:
        raise argparse.ArgumentTypeError(
            "'{0}' must be in format: BAND-VALUE:HTML-COLOR".format(s)
        )

    try:
        band_value = float(band_value)
    except ValueError:
        raise argparse.ArgumentTypeError(
            "'{0}' is not a valid number".format(band_value)
        )

    try:
        color = rgba.webcolor(html_color)
    except ValueError:
        raise argparse.ArgumentTypeError(
            "'{0}' is not a valid HTML color".format(html_color)
        )

    return band_value, color


def colorize_band_arg(s):
    """Validates --colorize-band"""
    try:
        result = int(s)
    except ValueError:
        raise argparse.ArgumentTypeError("invalid int value: '{0}'".format(s))
    if result <= 0:
        raise argparse.ArgumentTypeError(
            "'{0}' must be 1 or greater".format(s)
        )
    return result


def png8_arg(s):
    """Validates --png8"""
    if s is None:
        result = s
    else:
        try:
            result = int(s)
        except ValueError:
            raise argparse.ArgumentTypeError("invalid int value: '{0}'".format(s))
        if not 2 <= result <= 256:
            raise ValueError(
                'png8 must be between 2 and 256: {0!r}'.format(result)
            )
    return result


def parse_args(args):
    """Parses command-line `args`"""

    LatestMetadata = Metadata.latest()

    parser = argparse.ArgumentParser(
        description='Converts a GDAL-readable into an MBTiles file'
    )
    parser.add_argument('-v', '--verbose', action='count',
                        help='explain what is being done')

    group = parser.add_argument_group(title='Positional arguments')
    group.add_argument('INPUT', type=argparse.FileType('rb'), nargs='?',
                       default=sys.stdin,
                       help='GDAL-readable file.')
    group.add_argument('OUTPUT', type=argparse.FileType('wb'), nargs='?',
                       help='Output filename. Defaults to INPUT.mbtiles')

    group = parser.add_argument_group(title='MBTiles metadata arguments')
    group.add_argument('--name', default=None,
                       help=('Human-readable name of the tileset. '
                             'Defaults to INPUT'))
    group.add_argument('--description', default="",
                       help='Description of the layer. Defaults to ""')
    group.add_argument('--layer-type',
                       default=LatestMetadata.TYPES.OVERLAY,
                       choices=LatestMetadata.TYPES,
                       help='Type of layer. Defaults to "overlay"')
    group.add_argument('--version', default='1.0.0',
                       help='Version of the tileset. Defaults to "1.0.0"')
    group.add_argument('--format',
                       default=LatestMetadata.FORMATS.PNG,
                       choices=LatestMetadata.FORMATS,
                       help='Tile image format. Defaults to "png"')

    group = parser.add_argument_group(title='GDAL warp arguments')
    group.add_argument('--spatial-reference', type=int, default=3857,
                       help=('Destination EPSG spatial reference. '
                             'Defaults to 3857'))
    group.add_argument('--resampling',
                       default='near',
                       choices=list(RESAMPLING_METHODS.values()),
                       help=('Resampling algorithm for warping. '
                             'Defaults to "near" (nearest-neighbour)'))

    group = parser.add_argument_group(title='Rendering arguments')
    group.add_argument('--min-resolution', type=int, default=None,
                       help=('Minimum resolution/zoom level to render and slice. '
                             'Defaults to None (do not downsample)'))
    group.add_argument('--max-resolution', type=int, default=None,
                       help=('Maximum resolution/zoom level to render and slice. '
                             'Defaults to None (do not upsample)'))
    group.add_argument('--fill-borders',
                       action='store_const', const=True, default=True,
                       help=('Fill image to whole world with empty tiles. '
                             'Default.'))
    group.add_argument('--no-fill-borders', dest='fill_borders',
                       action='store_const', const=False,
                       help='Do not add borders to fill image.')
    group.add_argument('--zoom-offset', type=int, default=0,
                       metavar='N',
                       help=('Offset zoom level by N to fit unprojected '
                             'images to square maps. Defaults to 0.'))

    group = parser.add_argument_group(title='Coloring arguments')
    group.add_argument('--coloring', default=None,
                       choices=COLORING_METHODS,
                       help='Coloring algorithm.')
    group.add_argument('--color', dest='colors', action='append',
                       type=color_arg, metavar='BAND-VALUE:HTML-COLOR',
                       help=('Examples: --color="0:#ff00ff" --color=255:red'))
    group.add_argument('--colorize-band', metavar='COLORIZE-BAND',
                       type=colorize_band_arg, default=None,
                       help='Raster band to colorize. Defaults to 1')
    group.add_argument('--png8', default=None,
                       type=png8_arg,
                       help=('Quantizes 32-bit RGBA to 8-bit RGBA paletted '
                             'PNGs. If an integer, specifies number of '
                             'colors in palette between 2 and 256. '
                             'Default to False.'))

    args = parser.parse_args(args=args)

    # Guess at the OUTPUT based on the INPUT
    if args.OUTPUT is None:
        if args.INPUT == sys.stdin:
            args.OUTPUT = sys.stdout
        else:
            # Set default output name based on input name
            args.OUTPUT = open(
                os.path.splitext(args.INPUT.name)[0] + '.mbtiles',
                mode='wb'
            )

    if args.name is None:
        args.name = os.path.basename(args.INPUT.name)

    # Make sure that --color and --coloring match up
    if args.coloring is None and (args.colors or
                                  args.colorize_band is not None):
        parser.error('must provide --coloring')
    elif args.coloring is not None and not args.colors:
        parser.error('must provide at least one --color')

    # Transform choices into ColorBase classes
    args.coloring = coloring_arg(args.coloring)

    return args


@contextmanager
def input_output(inputfile, outputfile):
    tempfiles = []

    infile = inputfile
    if inputfile == sys.stdin:
        infile = NamedTemporaryFile()
        copyfileobj(inputfile, infile)
        infile.seek(0)
        tempfiles.append(infile)

    outfile = outputfile
    if outputfile == sys.stdout:
        outfile = NamedTemporaryFile()
        tempfiles.append(outfile)

    try:
        yield infile, outfile
        if outputfile == sys.stdout:
            copyfileobj(open(outfile.name, 'rb'), outputfile)
    finally:
        for f in tempfiles:
            f.close()


def main(args=None, use_logging=True):
    if args is None:
        args = sys.argv[1:]
    args = parse_args(args=args)

    if use_logging:
        configure_logging(args)

    # HACK: Import here, so that VIPS doesn't parse sys.argv!!!
    # In vimagemodule.cxx, SWIG_init actually does argument parsing
    from gdal2mbtiles.helpers import warp_mbtiles

    with input_output(inputfile=args.INPUT,
                      outputfile=args.OUTPUT) as (inputfile, outputfile):
        # MBTiles
        metadata = dict(
            description=args.description,
            format=args.format,
            name=args.name,
            type=args.layer_type,
            version=args.version,
        )

        # GDAL
        spatial_ref = SpatialReference.FromEPSG(args.spatial_reference)

        # Coloring
        if not args.coloring:
            colors = band = None
        else:
            colors = args.coloring(args.colors)
            band = args.colorize_band

        # PNG rendering
        pngdata = {'png8': args.png8}

        warp_mbtiles(inputfile=inputfile.name, outputfile=outputfile.name,
                     # MBTiles
                     metadata=metadata,
                     # GDAL
                     spatial_ref=spatial_ref, resampling=args.resampling,
                     # Rendering
                     min_resolution=args.min_resolution,
                     max_resolution=args.max_resolution,
                     fill_borders=args.fill_borders,
                     zoom_offset=args.zoom_offset,
                     pngdata=pngdata,
                     # Coloring
                     colors=colors, band=band)
        return 0


def configure_logging(args):
    if not args.verbose:
        return

    if args.verbose == 1:
        level = logging.INFO
        fmt = '%(message)s'
    else:
        level = logging.DEBUG
        fmt = '%(asctime)s %(module)s: %(message)s'

    logging.basicConfig(level=level, format=fmt,
                        datefmt='%Y-%m-%d %H:%M:%S')


if __name__ == '__main__':
    retcode = main(args=sys.argv[1:])
    sys.exit(retcode)
